src/main/java/org/zalando/nakadi/controller/CursorOperationsController.java

```diff
@@ -0,0 +1,196 @@
+    }
+
+    @ExceptionHandler(InvalidCursorOperation.class)
+    public ResponseEntity<?> invalidCursorOperation(final InvalidCursorOperation e,
+                                                          final NativeWebRequest request) {
+        LOG.debug("User provided invalid cursor for operation. Reason: " + e.getReason(), e);
+        return Responses.create(Problem.valueOf(MoreStatus.UNPROCESSABLE_ENTITY,
+                clientErrorMessage(e.getReason())), request);
+    }
+
+    private String clientErrorMessage(final InvalidCursorOperation.Reason reason) {
+        switch (reason) {
+            case INVERTED_TIMELINE_ORDER: return "Inverted timelines. Final cursor must correspond to a newer " +
+                    "timeline than initial cursor.";
+
+            case TIMELINE_NOT_FOUND: return "Timeline not found. It might happen in case the cursor refers to a " +
+                    "timeline that has already expired.";
+            case INVERTED_OFFSET_ORDER: return "Inverted offsets. Final cursor offsets must be newer than initial " +
+                    "cursor offsets";
+            case PARTITION_NOT_FOUND: return "Partition not found.";
+            case CURSORS_WITH_DIFFERENT_PARTITION: return "Cursors with different partition. Pairs of cursors should " +
+                    "have matching partitions.";
+            default: return unexpectedErrorReason(reason);
+        }
+    }
+
+    private void accessControl(final String eventTypeName, final Client client) {
+        final EventType eventType;
+        try {
+            eventType = eventTypeRepository.findByName(eventTypeName);
+            client.checkScopes(eventType.getReadScopes());
+        } catch (final InternalNakadiException e) {
+            throw new MyNakadiRuntimeException1("failed to get event type", e);
+        } catch (final NoSuchEventTypeException e) {
+            throw new NoEventTypeException(e.getMessage(), e);
+        }
+    }
+
+    private String unexpectedErrorReason(final InvalidCursorOperation.Reason reason) throws MyNakadiRuntimeException1 {
+        LOG.error("Unexpected invalid cursor operation reason " + reason);
+        throw new MyNakadiRuntimeException1();
+    }
+
+    private CursorLag toCursorLag(final NakadiCursorLag nakadiCursorLag) {
+        return new CursorLag(
+                nakadiCursorLag.getPartition(),
+                cursorConverter.convert(nakadiCursorLag.getFirstCursor()).getOffset(),
+                cursorConverter.convert(nakadiCursorLag.getLastCursor()).getOffset(),
+                nakadiCursorLag.getLag()
+        );
+    }
+
+    private Function<Cursor, NakadiCursor> toNakadiCursor(final String eventTypeName) {
+        return cursor -> {
+            try {
+                return cursorConverter.convert(eventTypeName, cursor);
+            } catch (final NakadiException | InvalidCursorException e) {
+                throw new CursorConversionException("problem converting cursors", e);
```

Comment:
What do you think of injecting this kind of method to `CursorConverter` (and later on - remove the old one)?
(Although maybe it's not the best ticket to do this modification)